File: 2022/04.py
# https://adventofcode.com/2022/day/4
# Wanted to use set methods for familiarisation, on the understanding it's not
# an efficient way to solve this problem when you can just compare bounds

import re

with open('./04data.txt') as datafile:
    input = datafile.readlines()

fullycontained = 0
overlap = 0

for line in input:
    bounds = [int(bound) for bound in re.split(',|-', line.strip())]
    set1 = set(range(bounds[0], bounds[1] + 1))
    set2 = set(range(bounds[2], bounds[3] + 1))
    if set1.issubset(set2) or set1.issuperset(set2):
        fullycontained += 1
    if not set1.isdisjoint(set2):
        overlap += 1

print(fullycontained)
print(overlap)

File: 2022/04.dart
// https://adventofcode.com/2022/day/4
// Solved in Python then reimplemented in Dart

import 'dart:io';
import 'dart:math';

void main() {
  final List<String> input = File('./04data.txt').readAsLinesSync();
  int fullyContained = 0;
  int overlap = 0;

  input.forEach((line) {
    final List<int> bounds = line
        .trim()
        .split(RegExp(',|-'))
        .map((e) => int.parse(e))
        .toList();

    if ((bounds[0] >= bounds [2] && bounds[1] <= bounds[3]) ||
        (bounds[2] >= bounds [0] && bounds[3] <= bounds[1])) {
      fullyContained++;
    }
    if (max(bounds[0], bounds[2]) <= min(bounds[1], bounds[3])) {
      overlap++;
    }
  });

  print(fullyContained);
  print(overlap);
}
